File: Games/RockPaperScissors/src/main/java/jokatu/game/games/rockpaperscissors/game/RockPaperScissorsGame.java
package jokatu.game.games.rockpaperscissors.game;

import jokatu.game.Game;
import jokatu.game.GameID;
import jokatu.game.player.StandardPlayer;
import jokatu.game.stage.GameOverStage;
import jokatu.game.stage.JoiningStage;
import jokatu.game.stage.machine.SequentialStageMachine;
import jokatu.game.status.StandardTextStatus;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

public class RockPaperScissorsGame extends Game<StandardPlayer> {

	public static final String ROCK_PAPER_SCISSORS = "Rock/paper/scissors";

	private final StandardTextStatus status = new StandardTextStatus();

	RockPaperScissorsGame(GameID identifier) {
		super(identifier);

		stageMachine = new SequentialStageMachine(Arrays.asList(
				() -> new JoiningStage<>(StandardPlayer.class, players, 2, status),
				() -> new InputStage(players.values(), status),
				() -> new GameOverStage(status)
		));

		status.observe(this::fireEvent);
	}

	@NotNull
	@Override
	public String getGameName() {
		return ROCK_PAPER_SCISSORS;
	}
}

File: Games/RockPaperScissors/src/main/kotlin/jokatu/game/games/rockpaperscissors/game/RockPaperScissorsGame.kt
package jokatu.game.games.rockpaperscissors.game

import jokatu.game.Game
import jokatu.game.GameID
import jokatu.game.player.StandardPlayer
import jokatu.game.stage.GameOverStage
import jokatu.game.stage.JoiningStage
import jokatu.game.stage.machine.SequentialStageMachine
import jokatu.game.status.StandardTextStatus

class RockPaperScissorsGame internal constructor(identifier: GameID) : Game<StandardPlayer>(identifier) {

	private val status = StandardTextStatus()

	override val gameName: String
		get() = ROCK_PAPER_SCISSORS

	init {

		stageMachine = SequentialStageMachine(
				{ JoiningStage(StandardPlayer::class.java, players, 2, status) },
				{ InputStage(players.values, status) },
				{ GameOverStage(status) }
		)

		status.observe({ this.fireEvent(it) })
	}

	companion object {
		const val ROCK_PAPER_SCISSORS = "Rock/paper/scissors"
	}
}
